lib/dry/types/map.rb

```diff
@@ -0,0 +1,110 @@
+module Dry
+  module Types
+    class Map < Definition
+
+      def initialize(keys: Undefined, values: Undefined, meta: EMPTY_HASH)
```

Comment:
Hi @fledman one more question, hope you do not mind, why we are using `Undefined` for default values for `keys` and `values`? Maybe we could use `Dry::Types::Any` and we avoid all the checking of `Undefined` in the class.

Also, I think it would be better for the `ast` having an empty map the result would be more idiomatic:
```ruby
[:map, [{:keys=>[:definition, [Object, {}]], :values=>[:definition, [Object, {}]]}, {}]])
```

instead of:

```ruby
[:map, [{}, {}]]
```

WDYT?